app/src/main/kotlin/at/bitfire/davdroid/settings/AccountSettingsMigrations.kt

```diff
@@ -58,9 +67,38 @@ class AccountSettingsMigrations @AssistedInject constructor(
         fun create(account: Account, accountSettings: AccountSettings): AccountSettingsMigrations
     }
 
-
     val accountManager: AccountManager = AccountManager.get(context)
 
+    /**
+     * With DAVx5 4.3.3 address book accounts now contain the collection ID as a unique identifier.
+     * We need to update the address book account names.
+     */
+    @Suppress("unused","FunctionName")
+    fun update_16_17() {
+        try {
+            context.contentResolver.acquireContentProviderClient(ContactsContract.AUTHORITY)
+        } catch (e: SecurityException) {
+            logger.log(Level.WARNING, "Missing permissions for contacts authority", e)
+            null
+        }.use { provider ->
+            if (provider == null) {
+                logger.warning("Couldn't connect to content provider of contacts authority")
+                return // Don't continue without provider
+            }
+            val service = serviceRepository.getByAccountAndType(account.name, Service.TYPE_CARDDAV)
+                ?: return
+            val oldAddressBookAccounts = accountRepository.getAddressBookAccounts().filter {
+                account.name == accountManager.getUserData(it, "real_account_name")
+            }
+            for (oldAddressBookAccount in oldAddressBookAccounts) {
+                val url = accountManager.getUserData(oldAddressBookAccount, LocalAddressBook.USER_DATA_URL)
+                val collection = collectionRepository.getByServiceAndUrl(service.id, url)
+                    ?: continue
+                val localAddressBook = localAddressBookFactory.create(oldAddressBookAccount, provider)
+                localAddressBook.update(collection, collection.forceReadOnly)
```

Comment:
Unfortunately, `forceReadOnly` is not always the same as `collection.forceReadOnly`.

There's a case when Managed DAVx5 has the global setting `Settings.FORCE_READ_ONLY_ADDRESSBOOKS` set – then `AddressBookSyncer.forceAllReadOnly` becomes `true`, and this will call `LocalAddressBook.update(forceReadOnly = true)` even when `info.forceReadOnly` (which indicates that the user has manually set the collection to _force read-only_) is `false`.

I don't know a good quick solution for that – ideally there would be a `LocalAddressBook.update()` that doesn't take / change `forceReadOnly`.

---

However some bigger issue: `LocalAddressBook.update()` doesn't seem to set/update `USER_DATA_COLLECTION_ID`. Renaming is fine, but `LocalAddressBook.findByCollection` still won't find the updated / renamed collection?

So I think it would probably be the best to always set the `USER_DATA_COLLECTION_ID` (and `USER_DATA_COLLECTION_URL` until https://github.com/bitfireAT/davx5/issues/603 is done) on `LocalAddressBook.update()`.

For `LocalAddressBook.create()` it's not a problem, because it uses `initialUserData`, which sets `USER_DATA_COLLECTION_ID` and `USER_DATA_COLLECTION_URL`.

Reply:
Good thing you remember such things! I have made the argument nullable, so it won't be changed at all if not provided.  
---
Oh wow. Yes that's important. I simply forgot that.